designs/v1beta1-full-changelist.md

```diff
@@ -0,0 +1,200 @@
+# Karpenter v1beta1 Full Change List
+
+This document formalizes the [v1beta1 laundry list](https://github.com/aws/karpenter/issues/1327) into a full change list of items that are coming in the migration from the `v1alpha5` APIs of Karpenter to `v1beta1`. This document purely describes the necessary changes and the rationale behind the changes. For the high-level overview of the API specs, view the [Karpenter v1beta1 Graduation](./v1beta1-api.md) design doc.
+
+### Update Kind/Group Naming
+
+As part of the bump to v1beta1, to allow the v1alpha5 APIs to exist alongside the v1beta1 APIs while users go through a migration process, the following renames are being proposed:
+
+1. API Group Renames
+    1. `karpenter.k8s.aws` → `compute.k8s.aws`
+2. Kind Renames
+    1. `Provisioner` → `NodePool`
+    2. `Machine` -> `NodeClaim`
+    3. `AWSNodeTemplate` → `NodeClass`
+
+We see the renames as opportunities to better align our API groups and kinds with upstream concepts as well as reducing confusion between other Kubernetes API concepts. Specifically, the word `Provisioner` (on its own) has become overloaded in Kubernetes, [particularly in the area of storage provisioning](https://kubernetes.io/docs/concepts/storage/storage-classes/#the-storageclass-resource). We want to get completely away from this naming, while also prefixing all of our kinds that apply to nodes with `Node` for better alignment and clarity across the project. 
+
+This gives the following naming to API types within the Karpenter project
+
+1. `karpenter.sh/NodePool`
+2. `karpenter.sh/NodeClaim`
+3. `compute.k8s.aws/NodeClass`
```

Comment:
If I have a cluster that uses AWS and (custom, not EKS Anywhere) on-prem nodes, what do I call the two different kinds? Are they both NodeClass in different API groups?

Reply:
Yes, we would expect that as we get more CloudProviders we would add more API groups but each API group would have the name `NodeClass`. This allows us to version each group separate from each other and give each of the cloudproviders some autonomy.

Reply:
Annoying for people who want to `kubectl get NodeClasses` though. It won't work how most people would like, I imagine.

Reply:
Could be solved with shortnames? Agree that, it's a touch awkward to have overlap, but I suspect it to be pretty rare. I'd hate the lose the simplicity in naming to account for a theoretical use case being more awkward. 

Frankly, kubectl should support overlapping kinds and return both!

Reply:
> kubectl should support overlapping kinds and return both

I feel like this starts to get complex when it comes to duplicate kind names from different API groups. Like, if you have two different nodeClasses from two different API groups with the same name, what happens when you do `kubectl edit nodeclass default`. It might start to get confusing to the user, at which point, a user just has to resort to specify the scoped resource name giving the full API group with `kubectl edit nodeclasses.compute.k8s.aws default` or some other cloudprovider to get the clarity they are desiring. 

I do doubt that we will have a lot of cross-cloud clusters using Karpenter (at least on the outset or maybe into the future) but the prospect of these naming conflicts does scare me a bit since they don't appear to be well-supported in K8s and fully based on the discovery order of the client.

Reply:
@sftim Did you have a recommendation on what to do here? I know projects like CAPI have resorted to just prefixing their kinds with the cloudprovider name like `AWSNodeClass`. Though I find that naming a little annoying as well.

Reply:
> AWSNodeClass

That's what I'd do; I think it'd match end user expectations.